### ISSUE TITLE
Log events from writes to stderr are no longer marked as errors via `isError`

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7903]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#7903|
|Type|release blocker: regression|
|Created|2015-05-19 13:45:46Z|
|Branch|https://github.com/twisted/twisted/tree/writestostderr-booyah-7903|

Compare:

```python
~/Projects/twisted/15.1$ python -c '                                                    
from twisted.python.log import startLogging, addObserver
from io import BytesIO
import sys, os
out = BytesIO()
startLogging(out)

def observe(event):
    observed.append(event)
observed = []
addObserver(observe)

print >>sys.stderr, b"Hello, world."
os.write(1, bytes(observed[0]["isError"]) + b"\n")
'
1
```

```python
~/Projects/twisted/15.2$ python -c '
from twisted.python.log import startLogging, addObserver
from io import BytesIO
import sys, os
out = BytesIO()
startLogging(out)

def observe(event):
    observed.append(event)
observed = []
addObserver(observe)

print >>sys.stderr, b"Hello, world."
os.write(1, bytes(observed[0]["isError"]) + b"\n")
'
0
```

<details><summary>Searchable metadata</summary>

```
trac-id__7903 7903
type__release_blocker__regression release blocker: regression
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_writestostderr_booyah_7903 branches/writestostderr-booyah-7903
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1432043146849517 1432043146849517
changetime__1432113785679011 1432113785679011
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_1"></a>|@hynek commented|
|-|-|

this is kind of ugly, an expedited 15.2.1 would be in order.  But I’m afraid we’re bound to run into more of such bugs in this release.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

(In [44867]) Branching to writestostderr-booyah-7903.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

I have no idea if this is barking up the wrong tree, but I've got a "fix".

Builders spun, please review.

I'll also commit to doing a quick 15.2.1 after this ticket is closed.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @wsanchez<br>@wsanchez set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Thanks.  Some stylistic comments:

  1. `x not in y` is a much nicer way of writing `not x in y` is 
  1. `test_levelNotChange` looks like it's about 4 different tests.
  1. I think it's okay to drop the attribution (to me in this ticket) in the test method.  If anything, you could list me as an author in the commit message if you want.
  1. There's no news fragment.

I'm not very familiar with the new logging system.  If I have more time later to dig in deeper I'll try to understand the behavioral implications for this change.  Or maybe someone who already knows the logging system can do a review.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_6"></a>|@wsanchez commented|
|-|-|

thanks for the quick fix, hawkowl.

I noticed exarkun's comment [#11338](https://github.com/twisted/twisted/issues/11338). I'm not sure it's a standards violation, but I agree it's more readable to say `x not in y`.

`test_levelNotChange` is testing a few combinations of the same potential problem.  I supposed that could be four tests, but I would seriously doubt the utility of such a change.

reviewed, modulo news file.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_7"></a>|@wsanchez commented|
|-|-|

Wait, caught one thing: in `_legacy.py`, don't set `isError=1` if `log_level` is `LogLevel.error`.

In the legacy system, `isError` is set when you call `err`, which is supposed to be called with an exception.  That's `LogLevel.critical` in the new system.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_8"></a>|@wsanchez commented|
|-|-|

…and never mind the above.

hawkowl points out that `twisted.python.log.logerr` emits at new-style `LogLevel.error`, which is correct, and a re-read of `err()` shows that it emits non-exception events with `isError=1`, so that's not abnormal.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set owner to @hawkowl<br>@glyph set status to `new`|
|-|-|

One additional note; the test docstring, "`Seemingly conflicting C{"log_level"} and C{"isError"} keys do not conflict.`", is unclear. "Conflict" is one of those words like "correct" and "broken" that seems descriptive at first but is actually pretty empty.  What would constitute a conflict?  What happens?

I would recommend something instead like "`If explicitly set, the C{isError} key will be preserved when forwarding from a new-style logging emitter to a legacy logging observer, regardless of log level.`"

And, to clarify wsanchez's feedback:

  - please optionally address these points, add a news file and merge.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44879]) Merge writestostderr-booyah-7903: Make legacy logging events from writing to stderr set isError correctly

Authors: exarkun, hawkowl
Reviewers: exarkun, wsanchez, glyph
Fixes: [#7903](https://github.com/twisted/twisted/issues/7903)
